### PR TITLE
[ Gather ] Dynamic Gather value propagation

### PR DESCRIPTION
**Pull request by jane-intel**

### CODE REVIEW

File: src/ngraph/pass/constant_folding_gather.cpp [hunk 1]
```diff
@@ -193,3 +194,62 @@ void pass::ConstantFolding::construct_constant_gather()
...
+        element::f32, Shape{2, 3, 4}, pattern::has_class<op::Concat>());
+    auto indices_label =
+        make_shared<pattern::op::Label>(element::i64, Shape{5}, pattern::has_class<op::Constant>());
+    auto axis_label =
+        make_shared<pattern::op::Label>(element::i64, Shape{1}, pattern::has_class<op::Constant>());
+    auto gather_v1 = make_shared<op::v1::Gather>(concat_label, indices_label, axis_label);
+
+    auto concat_gather_callback = [concat_label, indices_label, axis_label](pattern::Matcher& m) {
+        NGRAPH_DEBUG << "In callback for construct_constant_gather_with_subgraph against node = "
+                     << m.get_match_root()->get_name();
```
Comment (diyessi) [inline src/ngraph/pass/constant_folding_gather.cpp]: Skip the `->get_name()` and you'll get much more useful information.

Comment (mbencer) [reply src/ngraph/pass/constant_folding_gather.cpp]: done

File: src/ngraph/pass/constant_folding_gather.cpp [hunk 2]
```diff
@@ -193,3 +194,65 @@ void pass::ConstantFolding::construct_constant_gather()
...
+        if (concat_inputs.size() != shape_size(concat->get_shape()))
+            return false;
+
+        const int64_t rank = concat->get_shape()[0];
+        const int64_t raw_index = indices->cast_vector<int64_t>()[0];
+        const int64_t positive_index = raw_index < 0 ? rank + raw_index : raw_index;
+        const auto gathered =
+            concat_inputs[positive_index].get_source_output().get_node_shared_ptr();
+        if (!gathered->is_constant())
+            return false;
```
Comment (jane-intel) [inline src/ngraph/pass/constant_folding_gather.cpp]: Should we eliminate such Concat then?

File: test/constant_folding.cpp [hunk 3]
```diff
@@ -1596,6 +1596,70 @@ TEST(constant_folding, const_gather_v1_scalar)
     ASSERT_TRUE(test::all_close_f(values_out, values_expected, MIN_FLOAT_TOLERANCE_BITS));
 }
 
+#include "ngraph/pass/visualize_tree.hpp"
```
Comment (jane-intel) [inline test/constant_folding.cpp]: Please delete include

File: src/ngraph/pass/constant_folding_gather.cpp (hunk 2 above)
Comment (mbencer) [reply src/ngraph/pass/constant_folding_gather.cpp]: right, all block with creating constant was not needed

File: test/constant_folding.cpp (hunk 3 above)
Comment (mbencer) [reply test/constant_folding.cpp]: done

**Comment by mbencer** [comment]
> Overall looks good.
> 
> 1. I would appreciate `Concat` elimination mentioned above.
> 2. Testing for #1
> 3. Please add negative tests for cases when we are not allowed to fold

All points introduced

**Pull request merged**